### ISSUE TITLE
An improvement of definite integration algorithm

### ISSUE THREAD

**Issue by noteas** [fricas/fricas#118]
Dear FriCAS contributors,
Hi! I have found that for some functions without indefinite integral but with specific definite integral, the definite integral may be obtained from a formula called "interval reproduction formula":
$$\boxed{\int_a^bf(x)\mathrm dx=\int_a^bf(a+b-x)\mathrm dx}$$
Proof:
Let $x=a+b-t$, $\therefore t=a+b-x$.

$$\therefore \int_a^bf(x)\mathrm dx=\int_b^af(a+b-t)\mathrm d(a+b-t)=\int_b^af(a+b-t)\mathrm d(-t)=\int_a^bf(a+b-t)\mathrm dt=\int_a^bf(a+b-x)\mathrm dx\quad\mathrm{Q.E.D.}$$

However, as the reproduced integral has the same type of the original, it does no help to the calculation.
But this is what we get from the formula:
$$\boxed{\int_a^bf(x)\mathrm dx=\int_a^b\dfrac{f(x)+f(a+b-x)}{2}\mathrm dx}$$
And the point of this is that sth on the RHS may cancel out and make it integrable.
For an example: (interaction with FriCAS 1.3.8)
The integral $\int_{-\tfrac{\pi}{2}}^{\tfrac{\pi}{2}}\dfrac{3\mathrm e^x\sin^2x}{\mathrm e^x+1}\mathrm dx=\dfrac{3}{4}\pi$
```
(1) -> f(x)==3*exp(x)*sin(x)^2/(exp(x)+1)
                                                                   Type: Void
(2) -> integrate(f(x), x)
   Compiling function f with type Variable(x) -> Expression(Integer)
           x             2       %A
         ++  (- 3 cos(%A)  + 3)%e
   (2)   |   ---------------------- d%A
        ++            %A
                    %e   + 1
                                         Type: Union(Expression(Integer),...)
(3) -> integrate(f(x), x=-%pi/2..%pi/2, "noPole")

   (3)  "failed"
                                                Type: Union(fail: failed,...)
```
As we can see, the function has no indefinite integral and the definite integration fails.
But when we use the formula:
```
(4) -> integrate((f(x)+f(-x))/2, x=-%pi/2..%pi/2, "noPole")
   Compiling function f with type Polynomial(Integer) -> Expression(
      Integer)

        3 %pi
   (4)  -----
          4
                  Type: Union(f1: OrderedCompletion(Expression(Integer)),...)
```
We get the correct answer!
To sum up, this formula will be a effective method for definite integration, especially for those without indefinite integral.
My suggestion is when the normal approach of solving definite integral fails, try using the corollary of the interval reproduction formula and miracles may occur.
I hold a firm belief that this method can greatly improve FriCAS' ability of solving definite integrals.

Yours,
noteas

**Comment by ValZapod** [comment]
Well, ideally you just get the antiderivative: 3/8 (I E^(-2 I x) Hypergeometric2F1[-2 I,1,1-2 I,-E^x]-I E^(2 I x) Hypergeometric2F1[2 I,1,1+2 I,-E^x]+4 Log[1+E^x]-2 Sin[2 x]) 

And then you get 2.35619, which is 3/4 Pi.

**Comment by noteas** [comment]
> Well, ideally you just get the antiderivative: 3/8 (I E^(-2 I x) Hypergeometric2F1[-2 I,1,1-2 I,-E^x]-I E^(2 I x) Hypergeometric2F1[2 I,1,1+2 I,-E^x]+4 Log[1+E^x]-2 Sin[2 x])
> 
> And then you get 2.35619, which is 3/4 Pi.

However, that's not what I meant. Be aware of FriCAS fails to solve the integral directly, but succeeds solving the problem with the formula. So I consider the formula an effective tool to solve definite integrals for FriCAS.

Regards,
noteas
